### ISSUE TITLE
Implement clip_grad_norm for FSDP models

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#72548]
### 🚀 The feature, motivation and pitch

[Gradient clipping](https://deepai.org/machine-learning-glossary-and-terms/gradient-clipping#:~:text=Gradient%20clipping%20is%20a%20technique,usually%20in%20recurrent%20neural%20networks.&text=With%20gradient%20clipping%2C%20pre%2Ddetermined,down%20to%20match%20the%20norm.) is a popular technique to scale gradients and avoid exploding gradients issues in RNNs/very deep networks. PyTorch offers a util [torch.nn.utils.clip_grad_norm_](https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html) that enables users to clip gradients such that they collectively have a capped maximum norm.

For FSDP, since parameters are sharded across ranks, simply using clip_grad_norm_  would be incorrect as `model.parameters()` does not return the full set of unsharded parameters and thus gradient clipping would be applied incorrectly. This issue is to implement `fsdp_model.clip_grad_norm_` which handles the fact that the parameters are sharded appropriately.

Feel free to refer to FairScale implementation: https://fairscale.readthedocs.io/en/stable/_modules/fairscale/nn/data_parallel/fully_sharded_data_parallel.html#FullyShardedDataParallel.clip_grad_norm_. At a high level, we need to make sure we do the following:

1. Call `self._lazy_init` to ensure stream synchronization is done appropriately
2. Reduce local norm across all processes
3. Apply this reduced local norm on the model parameters
4. Ensure that this works with nested FSDP instances as well. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by daniellepintz** [comment]
Are you also planning to implement `clip_grad_value` for FSDP?

**Comment by pengyanghua** [comment]
> Are you also planning to implement `clip_grad_value` for FSDP?

Any plan to support this?
